---
title: "Untitled"
author: "Ismael Gutierrez-Fernandez"
date: "2025-04-28"
output: html_document
---

# This script should be run in parallel with the original DESeq2 script in order to perform the analysis without outlier(s)

```{r}
total_counts_for_DESeq2 <- total_counts_for_DESeq2[,-15] # Remove the column with the gene names (usually the first column)
head(total_counts_for_DESeq2)
dim(total_counts_for_DESeq2)
```

# Exploratory analysis without the outlier

```{r}
water_regime <- c("GAZUL_WW","GAZUL_WW","GAZUL_WW","GAZUL_WW","GAZUL_WW","GAZUL_LWD","GAZUL_LWD","GAZUL_LWD","GAZUL_LWD","GAZUL_LWD","GAZUL_TWD","GAZUL_TWD","GAZUL_TWD","GAZUL_TWD") # Create and name the vectors needed. Name the vectors in the same order as the replicates are
#CondVector <- rep(water_regime,each=5) # Indicate how many replicates are in each vector
#CondVector

counts <- round(total_counts_for_DESeq2) # Create a data set with the gene counts rounded
head(counts)

sampleTable <- data.frame(row.names = colnames(counts), conditions = as.factor(water_regime)) # Create the data frame for the matrix that will be used for the analysis. Define the CondVector previously created as the conditions factor for the analysis
sampleTable
head(sampleTable) 

dds <- DESeqDataSetFromMatrix(countData = counts, colData = sampleTable, design=~conditions) 
dds

dds$conditions <- relevel(dds$conditions, "GAZUL_WW") # Define which of the conditions is the reference
dim(dds)

# PCA plot of the dds matrix 
vsd <- vst(dds, blind=FALSE)
pcaData <- plotPCA(vsd, intgroup=c("conditions"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
nudge <- position_nudge(y = 5)

plot <- ggplot(pcaData, aes(x = PC1, y = PC2, color = conditions, label = name)) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(size = 5) +
  theme_bw() +
  scale_color_manual(values = c("#6ac7fc","#ebd834", "#d10a0a")) +
  ggtitle("GAZUL under two water deficits ") +
  geom_label(aes(label = name), position = nudge, size = 3, size.unit = "mm")
print(plot)

```

# RUN THE DESEQ2 

#### 2.1.1. Preparation of the matrix for each water deficit treatment contrast with DESeq2. The contrast are done by indpendently analysis for each comparison.

##### OPTION 1: Applying the specific filter for LWD and TWD respectively, instead of the common TPM > 0.5 filter use for the 3 conditions matrix
```{r}
# 2.1.1.1. Create the data set for LWD contrast applying the tpm > 0.5 table
LWD_merge_tpm <- list(GAZUL_WW_genes_0.5tpm, GAZUL_LWD_genes_0.5tpm) %>% reduce(full_join, by = "gene")
row.names(LWD_merge_tpm) <- LWD_merge_tpm$gene
LWD_merge_tpm <- LWD_merge_tpm[!grepl("LC",LWD_merge_tpm$gene), ]
dim(LWD_merge_tpm)

GAZUL_WW_by_LWD.counts_0.5tpm <- GAZUL_WW.count.data.2[rownames(GAZUL_WW.count.data.2) %in% LWD_merge_tpm$gene,]
GAZUL_WW_by_LWD.counts_0.5tpm$gene <- rownames(GAZUL_WW_by_LWD.counts_0.5tpm)

GAZUL_LWD.counts_0.5tpm <- GAZUL_LWD.count.data.2[rownames(GAZUL_LWD.count.data.2) %in% LWD_merge_tpm$gene,]
GAZUL_LWD.counts_0.5tpm$gene <- rownames(GAZUL_LWD.counts_0.5tpm)

counts_for_DESeq2_LWD <- list(GAZUL_WW_by_LWD.counts_0.5tpm, GAZUL_LWD.counts_0.5tpm) %>% reduce(full_join, by = "gene")
rownames(counts_for_DESeq2_LWD) <- counts_for_DESeq2_LWD$gene # Put the names of the genes in the merged table as the rownames for the data set
counts_for_DESeq2_LWD <- counts_for_DESeq2_LWD[,-6] # Remove the column with the gene names (usually the first column)
head(counts_for_DESeq2_LWD)
dim(counts_for_DESeq2_LWD)


# 2.1.1.1. Create the data set for TWD contrast applying the tpm > 0.5 table
TWD_merge_tpm <- list(GAZUL_WW_genes_0.5tpm, GAZUL_TWD_genes_0.5tpm) %>% reduce(full_join, by = "gene")
row.names(TWD_merge_tpm) <- TWD_merge_tpm$gene
TWD_merge_tpm <- TWD_merge_tpm[!grepl("LC",TWD_merge_tpm$gene), ]
dim(TWD_merge_tpm)

GAZUL_WW_by_TWD.counts_0.5tpm <- GAZUL_WW.count.data.2[rownames(GAZUL_WW.count.data.2) %in% TWD_merge_tpm$gene,]
GAZUL_WW_by_TWD.counts_0.5tpm$gene <- rownames(GAZUL_WW_by_TWD.counts_0.5tpm)

GAZUL_TWD.counts_0.5tpm <- GAZUL_TWD.count.data.2[rownames(GAZUL_TWD.count.data.2) %in% TWD_merge_tpm$gene,]
GAZUL_TWD.counts_0.5tpm$gene <- rownames(GAZUL_TWD.counts_0.5tpm)

counts_for_DESeq2_TWD <- list(GAZUL_WW.counts_0.5tpm, GAZUL_TWD.counts_0.5tpm) %>% reduce(full_join, by = "gene") %>% na.omit()
rownames(counts_for_DESeq2_TWD) <- counts_for_DESeq2_TWD$gene # Put the names of the genes in the merged table as the rownames for the data set
counts_for_DESeq2_TWD <- counts_for_DESeq2_TWD[,-6]# Remove the column with the gene names (usually the first column)
counts_for_DESeq2_TWD <- counts_for_DESeq2_TWD[,-10] # Remove the column outlier
head(counts_for_DESeq2_TWD)
dim(counts_for_DESeq2_TWD)

```



```{r}
# 2.1.2.1. Define the groups (factors) for the analysis.

water_deficit <- c("_TWD")

# 2.1.2.2. Get data for each point control vs treatment. A loop is created based on vector "water_deficit"

for (i in water_deficit) {

water_regime <- c(paste0("GAZUL_WW"),paste0("GAZUL_WW"),paste0("GAZUL_WW"),paste0("GAZUL_WW"),paste0("GAZUL_WW"),paste0("GAZUL_TWD"),paste0("GAZUL_TWD"),paste0("GAZUL_TWD"),paste0("GAZUL_TWD")) # Create and name the vectors needed. Name the vectors in the same order as the replicates are
#CondVector <- rep(water_regime,each=5) # Indicate how many replicates are in each vector
#CondVector

counts <- round(get(paste0("counts_for_DESeq2_TWD"))) # Create a data set with the gene counts rounded
head(counts)

sampleTable <- data.frame(row.names = colnames(counts), conditions = as.factor(water_regime)) # Create the data frame for the matrix that will be used for the analysis. Define the CondVector previously created as the conditions factor for the analysis
sampleTable
head(sampleTable) 

dds <- DESeqDataSetFromMatrix(countData = counts, colData = sampleTable, design=~conditions) 
dds

dds$conditions <- relevel(dds$conditions, "GAZUL_WW") # Define which of the conditions is the reference
dim(dds)

# PCA plot of the dds matrix 
vsd <- vst(dds, blind=FALSE) # Transformation of the variance of the standard deviation
pcaData <- plotPCA(vsd, intgroup=c("conditions"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
nudge <- position_nudge(y = 5)

plot <- ggplot(pcaData, aes(x = PC1, y = PC2, color = conditions, label = name)) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  coord_fixed() +
  geom_point(size = 5) +
  theme_bw() +
  scale_color_manual(values = c("darkgrey", "black")) +
  ggtitle("GAZUL under one treatment") +
  geom_label(aes(label = name), position = nudge, size = 3, size.unit = "mm")
print(plot)

# 2.2.1. Run the function

dds_analysis <- DESeq(dds)

dds_res <- results(dds_analysis, contrast = c("conditions",paste0("GAZUL_TWD"), paste0("GAZUL_WW")))

# Sort results on padj
ordered_res <- dds_res[order(dds_res$padj),] # Reorder based on the padj value
head(ordered_res)
tail(ordered_res)
ordered_res_na.rm <- na.omit(ordered_res) # Remove the NA gaps
head(ordered_res_na.rm)
tail(ordered_res_na.rm)

# Output ordered_res to csv
write.csv(ordered_res_na.rm, file = paste0(out_dir_DESeq2,"GAZUL_WW_TWD_results_all_no_outliers.csv")) # Save all the results for the vulcano plots

write.csv(ordered_res_na.rm[ordered_res_na.rm$padj<0.05,],file = paste0(out_dir_DESeq2,"GAZUL_WW_TWD_results_no_outiers.csv")) # Save the results whose padj is lower than 0.05

assign(paste("GAZUL_WW_TWD.na.rm",sep=""), ordered_res_na.rm[ordered_res_na.rm$padj<0.05,]) # Important the name given here is used below

print(paste("GAZUL_WW_TWD.na.rm",sep=""))
dim(assign(paste("GAZUL_WW_TWD.na.rm",sep=""), ordered_res_na.rm[ordered_res_na.rm$padj<0.05,]))

resLFC <- lfcShrink(dds_analysis, coef=paste0("conditions_GAZUL_TWD_vs_GAZUL_WW"), type="ashr")
print(plotMA(resLFC, ylim=c(-2,2)))

}
##############################################################

DE_genes <- data.frame(treatment = character(),
                       DE_genes_0.05 = numeric(), upreg_0.05 = numeric(), downreg_0.05 = numeric(),
                       DE_genes_0.01 = numeric(), upreg_0.01 = numeric(), downreg_0.01 = numeric(),
                       DE_genes_0.001 = numeric(), upreg_0.001 = numeric(), downreg_0.001 = numeric())

DE_genes

for (i in water_deficit) {

head(get(paste("GAZUL_WW_TWD.na.rm",sep="")))
my_data <- get(paste("GAZUL_WW_TWD.na.rm",sep="")) # Get the data saved above
head(my_data)

# Extract the data from the DESeq2 analysis. The range of p-value are those who where in the "DE_genes" data frame
DE_genes_0.05 <- nrow(my_data) 
upreg_0.05 <- nrow(my_data[my_data$log2FoldChange > 1,])
downreg_0.05 <- nrow(my_data[my_data$log2FoldChange < -1,])
DE_genes_0.01 <- nrow(my_data[my_data$padj < 0.01,])
upreg_0.01 <- nrow(my_data[my_data$padj < 0.01 & my_data$log2FoldChange > 1,])
downreg_0.01 <- nrow(my_data[my_data$padj < 0.01 & my_data$log2FoldChange < -1,])
DE_genes_0.001 <- nrow(my_data[my_data$padj < 0.001,])
upreg_0.001 <- nrow(my_data[my_data$padj < 0.001 & my_data$log2FoldChange > 1,])
downreg_0.001 <- nrow(my_data[my_data$padj < 0.001 & my_data$log2FoldChange < -1,])

# Fill the data frame with the data extracted
DE_genes <- rbind(DE_genes, list(treatment = as.character(i),
                                 DE_genes_0.05 = DE_genes_0.05, upreg_0.05 = upreg_0.05, downreg_0.05 = downreg_0.05,
                                 DE_genes_0.01 = DE_genes_0.01, upreg_0.01 = upreg_0.01, downreg_0.01 = downreg_0.01,
                                 DE_genes_0.001 = DE_genes_0.001, upreg_0.001 = upreg_0.001, downreg_0.001 = downreg_0.001))
}
DE_genes

colnames(DE_genes) <- c("treatment", "DE_genes_0.05", "upreg2fold_0.05", "downreg2fold_0.05", "DE_genes_0.01", "upreg2fold_0.01", "downreg2fold_0.01", "DE_genes_0.001", "upreg2fold_0.001", "downreg2fold_0.001") # Names of the columns in the data frame. Should match the names defined above to avoid confusions

write.csv(DE_genes, file = paste0(out_dir_DESeq2,"GAZUL","_DE_genes_results_no_outiers.csv")) # Save the data frame/table created with the proper name in .csv format
```

### 3.2. GO term enrichment. Manual: https://bioinformatics-core-shared-training.github.io/cruk-summer-school-2020/RNAseq/extended_html/06_Gene_set_testing.html

```{r}
# 3.2.1. Create the data frame for the output

GO_enriched <- data.frame(category = character(), over_represented_pvalue =  numeric(), under_represented_pvalue = numeric(), numDEInCat = numeric(), numInCat = numeric(), term = character(), ontology = character(), over_rep_padj = numeric(), treatment_threshold = character())

head(GO_enriched)

# 3.2.2. Do the GO enrichment for each time point. The time_list object was defined in DESeq2 #2.2.2. chunk

for (i in water_deficit) {
  
  head(get(paste("GAZUL_WW",i,".na.rm",sep = ""))) # Define in in DESeq2 #2.2.2. chunk, for each time point in the item list
  my_data <- get(paste("GAZUL_WW",i,".na.rm",sep = "")) # Get the data from DESeq2 analysis
  head(my_data)
  
  # Create a sublist of items (j) each item is the vector with the differential expressed genes inside the range (based on the padj of DESeq2). We can choose the ranges that are more interesting
  upreg2fold_0.05 <- (my_data[my_data$padj < 0.05 & my_data$log2FoldChange > 1,])
  downreg2fold_0.05 <- (my_data[my_data$padj < 0.05 & my_data$log2FoldChange < -1,])
  upreg2fold_0.01 <- (my_data[my_data$padj < 0.01 & my_data$log2FoldChange > 1,])
  downreg2fold_0.01 <- (my_data[my_data$padj < 0.01 & my_data$log2FoldChange < -1,])
  upreg2fold_0.001 <- (my_data[my_data$padj < 0.001 & my_data$log2FoldChange > 1,])
  downreg2fold_0.001 <- (my_data[my_data$padj < 0.001 & my_data$log2FoldChange < -1,])
  
  for (j in c("upreg2fold_0.05","downreg2fold_0.05","upreg2fold_0.01","downreg2fold_0.01","upreg2fold_0.001","downreg2fold_0.001")) { # Indicate the name of the (j) items as they were defined just above
    genes_for_GO <- (get(j))
    head(genes_for_GO)
    
    # Now do GO stats analysis on the genes expressed in each pattern compared to all genes expressed
    # Create a named binary vector for genes where one means differently expressed and 0 means not differentially expressed
    de.genes <- rownames(genes_for_GO)
    gene.vector= as.integer(assayed.genes%in%de.genes)
    names(gene.vector)=assayed.genes
    head(gene.vector)
    
    # Carry out the GOseq analysis
    
    pwf = nullp(gene.vector, bias.data = gene.lens, plot.fit = TRUE)
    GO.wall = goseq(pwf,gene2cat = all_go_subset)
    # This gave table with p-values. Now correct for multiple testing using FDR
    enriched.GO = GO.wall$category[p.adjust(GO.wall$over_represented_pvalue, method = "BH")<.05]
    head(enriched.GO)
    
    # Add new column with over represented GO terms padj
    GO.wall$over_rep_padj = p.adjust(GO.wall$over_represented_pvalue, method = "BH")
    write.table(GO.wall[GO.wall$over_rep_padj<0.05,], file = paste0(out_dir_GO,"GAZUL",i,"_vs_WW_results_",j,"_GOseq_no_outliers.tsv", sep = ""), sep = "\t", quote = FALSE, col.names = TRUE, row.names = F)
    GO_enriched_timepoint <- GO.wall[GO.wall$over_rep_padj <0.05 & GO.wall$ontology == "BP",]
    head(GO_enriched_timepoint)
    
    # if no enriched GO terms don't add to dataframe
    
    if(nrow(GO_enriched_timepoint)>0) {
    
    GO_enriched_timepoint$treatment_threshold <- paste0(i,"_treatment_",j) # Warning: the way to call each category in this column "timepoint_threshold" will be important for one step ahead
    GO_enriched <- rbind(GO_enriched,GO_enriched_timepoint)
    
    }
}
}
```
